### PR TITLE
Fix #13687: Move to latest App Services MSI API version

### LINKED ISSUE #13687: Move to latest App Services MSI API version
- Move Azure.Identity to the latest App Services MSI API version.
- Update /sdk/e2e Azure Functions date parsing tests

### PR DESCRIPTION
**Pull request by AlexanderSher**

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs
Comment (schaabs) [inline sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs:84]: Should this change be in a separate PR? Are there not other libraries using the explicit methods you're removing here?

**Review by schaabs — CHANGES_REQUESTED** [review]
I like how you've separated the ManagedIdentityClient into implementations for each type of managed identity endpoint we're dealing with. I think this will scale better as we continue to support more managed identity hosts with unique protocols. 

However, I think this could be cleaned up even further if you increase the scope of the abstraction. Currently the abstraction `IAuthRequestBuilder` is only responsible for creating the request, however, there are differences between the managed identity protocols which go beyond differences in the request. What if the abstraction was something like `IManagedIdentityProvider` with methods like `IsAvailable` and `GetToken` so the entire authentication request flow can be handled by the implementation. Thoughts?

File: sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs:84]: Text search over our codebase doesn't show any other usages of this method.

File: sdk/identity/Azure.Identity/src/IManagedIdentitySource.cs
Comment (AlexanderSher) [inline sdk/identity/Azure.Identity/src/IManagedIdentitySource.cs:11]: This is as good as I was able to make it. There are still code pieces that are mostly common, like extracting json properties, so I've left them in the `ManagedIdentityClient`

**Review by schaabs — APPROVED** [review]
Approving from my end. Please wait for @chlowell  to approve before merging.

File: sdk/identity/Azure.Identity/src/ManagedIdentityCredential.cs
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/ManagedIdentityCredential.cs:74]:
```suggestion
        /// Obtains an <see cref="AccessToken"/> from the Managed Identity service if available. This method is called by Azure SDK clients. It isn't intended for use in application code.
```

File: sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs:128]: Looks like the credential will send another request the next time a token is requested. Is this a change from the prior behavior?

**Review by chlowell — APPROVED** [review]
Managed identity bits look good to me.

File: sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs:128]: No, error message is stored in `_identityUnavailableErrorMessage` and will be used later to throw exception: https://github.com/Azure/azure-sdk-for-net/pull/14770/files/807bb81803ec86e460538d8c76d101ff57f68c9f#diff-7bbc16093e3640db68eb0921760088e1R84

**Pull request merged**
